## Data-Driven Analysis of eBay Classifieds Used Cars

This project aims to work with a dataset of used cars from eBay Kleinanzeigen, German [classifieds](https://en.wikipedia.org/wiki/Classified_advertising) section of the eBay website. The data was [scraped](https://en.wikipedia.org/wiki/Web_scraping) and uploaded by user [Orgesleka](https://www.kaggle.com/orgesleka) on Kaggle. While the original dataset is not available on Kaggle, you can download it from [here](https://data.world/data-society/used-cars-data).

Project goal:

In this project, we will go through the dataset to clean it and a preliminary analysis. We try to bring out informative knowledge about the second-hand dull car market of Germany.


## Introduction
Let's begin by importing the necessary libraries and then loading the dataset:

In [3]:
import numpy as np
import pandas as pd

# Load the dataset into a pandas DataFrame
# Utilize 'latin1' encoding to manage special characters in the dataset

autos = pd.read_csv('autos.csv', encoding = 'latin1')

## Data Exploration

To get started, lets explore the dataset using `DataFrame. info()` and `DataFrame. head()` methods. Using these methods we are able to look into the construction and content of autos DataFrame:

In [5]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

The dataset consists of 50,000 entries and 20 columns.

Most columns have complete data, except for a few:
* vehicleType (44,905 non-null)
* gearbox (47,320 non-null)
* model (47,242 non-null)
* fuelType (45,518 non-null)
* notRepairedDamage (40,171 non-null)

In [7]:
# autos.head(): shows the initial rows of the DataFrame, offering a preview of the data and the first few records.
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Observations:
* Data Columns: The dataset includes columns such as dateCrawled, name, seller, offerType, price, vehicleType, yearOfRegistration, gearbox, powerPS, model, odometer, monthOfRegistration, fuelType, brand, notRepairedDamage, dateCreated, nrOfPictures, postalCode, and lastSeen.
* Price: Prices are recorded in dollars and may include formatting issues that need cleaning.
* Vehicle Information: Details about the vehicle type, model, year of registration, gearbox, and power are provided.
* Missing Values: Some columns, such as vehicleType, gearbox, model, fuelType, and notRepairedDamage, have missing values.
* Date Columns: Dates are provided in several columns, capturing when the ad was crawled, created, and last seen.

## Cleaning Column Names
The column names are in `CamelCase` rather than Python's preferred `snake_case` format, so we can't simply replace spaces with underscores. We'll convert the column names from camelCase to snake_case and adjust some of them for clarity using the data dictionary.

In [10]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest', 'vehicle_type', 'registration_year', 'gear_box', 'power_ps', 'model', 'odometer', 
                 'registration_month', 'fuel_type', 'brand', 'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code', 'last_seen']


In [11]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

    gear_box  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## Initial Exploration and Cleaning
Let's start some very basic data exploration to identify some areas that need cleaning. We'll look at:
* Text columns where all or close to all the values are the same.
* Examples of numeric data, stored as text, that may be cleaned and converted to proper numeric format.

In [13]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year gear_box      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

After inspecting the data, we find the following:

* Most values entered in the columns `seller` and `offer_type` are one of two values: `privat` for seller and `Angebot` for offer_type.
* With this, the data in the `price` and `odometer` columns have to be represented as integers.
* The price column needs more research as the most frequently occurring value is `0`.

In [15]:
# Remove $ sign character and commas, then change the data type to float:
autos['price'] = autos['price'].str.replace('$', '').str.replace(',', '')
autos['price'] = autos['price'].astype('float')

# Verify changes
print(autos["price"].head())
print(autos["price"].dtype)

0    5000.0
1    8500.0
2    8990.0
3    4350.0
4    1350.0
Name: price, dtype: float64
float64


In [16]:
# Remove any non-numeric characters, then change the data type to float and rename the column:
autos.rename(columns = {'odometer': 'odometer_km'}, inplace=True)

autos['odometer_km'] = autos['odometer_km'].str.replace('km', '').str.replace(',', '')
autos['odometer_km'] = autos['odometer_km'].astype('float')

# Verify changes
print(autos["odometer_km"].head())
print(autos["odometer_km"].dtype)

0    150000.0
1    150000.0
2     70000.0
3     70000.0
4    150000.0
Name: odometer_km, dtype: float64
float64


## Exploring the Odometer and Price Columns

We earlier noticed mistakes in the price and odometer_km columns. Let's look closely at these columns to highlight and correct any problems with the columns.

We will carefully go through these columns to:
 
* Detect any misentered or outlier values that may bias our analysis.
* Check that your data is well formatted and contains real-world values.

In [18]:
# Examine "odometer_km" unique values
print(autos['odometer_km'].unique().shape)
print('\n')
print(autos['odometer_km'].describe())
print('\n')
autos['odometer_km'].value_counts()

(13,)


count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64




odometer_km
150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
70000.0      1230
60000.0      1164
50000.0      1027
5000.0        967
40000.0       819
30000.0       789
20000.0       784
10000.0       264
Name: count, dtype: int64

The low number of mileage categories, combined with the rounded values, suggests sellers likely selected from a pre-populated list of mileages.

In [20]:
# Examine "odometer_km" unique values
print(autos['price'].unique().shape)
print('\n')
print(autos['price'].describe())
print('\n')
autos['price'].value_counts()

(2357,)


count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64




price
0.0        1421
500.0       781
1500.0      734
2500.0      643
1000.0      639
           ... 
414.0         1
79933.0       1
5198.0        1
18890.0       1
16995.0       1
Name: count, Length: 2357, dtype: int64

The prices of cars range from 0 to 100 million, and with 2,357 rounded prices, it seems yes, sellers do tend to round their prices. There are also 1,421 instances where the price is 0, which we may want to clean.

In [22]:
# Let`s check the outliers
print(autos['price'].value_counts().sort_index().head(15))
print('\n')
print(autos['price'].value_counts().sort_index(ascending = False).head(15))

price
0.0     1421
1.0      156
2.0        3
3.0        1
5.0        2
8.0        1
9.0        1
10.0       7
11.0       2
12.0       3
13.0       2
14.0       1
15.0       2
17.0       3
18.0       1
Name: count, dtype: int64


price
99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
Name: count, dtype: int64


Well, it is an auction platforme, so probably all the auctions that the sellers place start at `$1`. We see how prices increase gradually up to `$350000`, followed by a huge rise to unrealistically high prices. So we will remove the prices over $350,000 also.

In [24]:
autos = autos[autos["price"].between(1, 350000)]

autos.describe(include='all')

date_crawled         name  seller offer_type          price  \
count                 48565        48565   48565      48565   48565.000000   
unique                46882        37470       2          1            NaN   
top     2016-03-21 20:37:19  Ford_Fiesta  privat    Angebot            NaN   
freq                      3           76   48564      48565            NaN   
mean                    NaN          NaN     NaN        NaN    5888.935591   
std                     NaN          NaN     NaN        NaN    9059.854754   
min                     NaN          NaN     NaN        NaN       1.000000   
25%                     NaN          NaN     NaN        NaN    1200.000000   
50%                     NaN          NaN     NaN        NaN    3000.000000   
75%                     NaN          NaN     NaN        NaN    7490.000000   
max                     NaN          NaN     NaN        NaN  350000.000000   

       abtest vehicle_type  registration_year gear_box      power_ps  model  \
count   48565        43979       48565.000000    46222  48565.000000  46107   
unique      2            8                NaN        2           NaN    245   
top      test    limousine                NaN  manuell           NaN   golf   
freq    25019        12598                NaN    36102           NaN   3900   
mean      NaN          NaN        2004.755421      NaN    117.197158    NaN   
std       NaN          NaN          88.643887      NaN    200.649618    NaN   
min       NaN          NaN        1000.000000      NaN      0.000000    NaN   
25%       NaN          NaN        1999.000000      NaN     71.000000    NaN   
50%       NaN          NaN        2004.000000      NaN    107.000000    NaN   
75%       NaN          NaN        2008.000000      NaN    150.000000    NaN   
max       NaN          NaN        9999.000000      NaN  17700.000000    NaN   

          odometer_km  registration_month fuel_type       brand  \
count    48565.000000        48565.000000     44535       48565   
unique            NaN                 NaN         7          40   
top               NaN                 NaN    benzin  volkswagen   
freq              NaN                 NaN     29368       10336   
mean    125770.101925            5.782251       NaN         NaN   
std      39788.636804            3.685595       NaN         NaN   
min       5000.000000            0.000000       NaN         NaN   
25%     125000.000000            3.000000       NaN         NaN   
50%     150000.000000            6.000000       NaN         NaN   
75%     150000.000000            9.000000       NaN         NaN   
max     150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nr_of_pictures   postal_code  \
count              39464                48565         48565.0  48565.000000   
unique                 2                   76             NaN           NaN   
top                 nein  2016-04-03 00:00:00             NaN           NaN   
freq               34775                 1887             NaN           NaN   
mean                 NaN                  NaN             0.0  50975.745207   
std                  NaN                  NaN             0.0  25746.968398   
min                  NaN                  NaN             0.0   1067.000000   
25%                  NaN                  NaN             0.0  30657.000000   
50%                  NaN                  NaN             0.0  49716.000000   
75%                  NaN                  NaN             0.0  71665.000000   
max                  NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 48565  
unique                38474  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

The `odometer_km` column had no outliers.

For the `price` column, we identified outliers using limits between `$1` and `$350000`.
The `Mean` after making that change is `$5888.93`

Car prices will account more for relatively realistic value, thus improving the reliability of our analysis.

olumns.

## Exploring the Date Columns

There are a few columns with dates in them: 
* date_crawled
* registration_year
* registration_month
* ad_created
* last_seen

Of these, all of the non-registration dates are stored as strings by pandas. Let's start by taking a look at these columns.

In [28]:
# Calculate the distribution of values in "date_crawled" as percentages
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

date_crawled
2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: proportion, dtype: float64

It seems to have been crawled on a daily basis within the two-month period starting in March of 2016. The listings collected each day are relatively consistent with regard to the distribution.

In [30]:
# Calculate the distribution of values in "ad_created" as percentages
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

ad_created
2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: proportion, Length: 76, dtype: float64

Most `ad_created` dates are no older than 1-2 months from the listing date, while some are extremely older — up to 9 months. There is a clear increase in the number of ads from 2016-03-05 to 0.036858.

In [32]:
# Calculate the distribution of values in "last_seen" as percentages
print(autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index())

last_seen
2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: proportion, dtype: float64


It logged the last date it observed each listing, which allows us to infer when a listing was removed, likely because the car had been sold.

The final three days all have an inordinately high number of `last_seen` values, 6-10 times greater than on earlier days. This is more likely due to the end of the crawling period rather than a sudden spike in sales.

## Dealing with Incorrect Registration Year Data

In [35]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The above output shows at a minimum `1000` and a maximum of `9999` for the column containing registration_year. These values are obviously unrealistic:
* Because a car can't be first registered after the listing was seen, any vehicle with a registration year above `2016` is definitely inaccurate.
* The minimum value represents a time before motor vehicles were even invented.

We will have to define an acceptable range for the year of vehicle registration, and all values outside of this range need to be removed. This is what will help in keeping our dataset valid and reliable.

We consider a certain range for the year of registration to be within the interval from `1900` to the current year `2016`. We will remove values beyond these bounds and compute below the distribution of the remaining values.

In [37]:
# Removing values 
autos = autos[autos["registration_year"].between(1900, 2016)]

In [38]:
print(autos["registration_year"].value_counts(normalize=True).sort_index())

registration_year
1910    0.000107
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000043
          ...   
2012    0.028063
2013    0.017202
2014    0.014203
2015    0.008397
2016    0.026135
Name: proportion, Length: 78, dtype: float64


In [39]:
autos['registration_year'].describe()

count    46681.000000
mean      2002.910756
std          7.185103
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

The cleaned `registration_year` data now makes a more realistic and meaningful distribution. Most of the cars in this dataset are primarily from the late 1990s to the early 2000s, with some listings dating back to the middle of the 20th century. This gives an improved picture of what age range most of the cars being listed fall into, which is important in gaining a better idea of where the data falls when performing further analysis.

## Exploring Price by Brand

In [42]:
top_brands = autos['brand'].value_counts().head(20)
print(top_brands)

brand
volkswagen        9862
bmw               5137
opel              5022
mercedes_benz     4503
audi              4041
ford              3263
renault           2201
peugeot           1393
fiat              1197
seat               853
skoda              766
nissan             713
mazda              709
smart              661
citroen            654
toyota             593
hyundai            468
sonstige_autos     458
volvo              427
mini               409
Name: count, dtype: int64


In [43]:
top_brand_by_price = {}

for brand in top_brands.index:
    mean_price = autos[autos["brand"] == brand]["price"].mean()
    top_brand_by_price[brand] = mean_price.round(2)

print(top_brand_by_price)

{'volkswagen': 5402.41, 'bmw': 8332.82, 'opel': 2975.24, 'mercedes_benz': 8628.45, 'audi': 9336.69, 'ford': 3749.47, 'renault': 2474.86, 'peugeot': 3094.02, 'fiat': 2813.75, 'seat': 4397.23, 'skoda': 6368.0, 'nissan': 4743.4, 'mazda': 4112.6, 'smart': 3580.22, 'citroen': 3779.14, 'toyota': 5167.09, 'hyundai': 5365.25, 'sonstige_autos': 12338.55, 'volvo': 4946.5, 'mini': 10613.46}


**Observations:**

* The premium brands are Audi, Mercedes-Benz, and BMW. All of these aforementioned brands have the highest average prices, thus positioning them in this premium segment of the market. Audi and Mercedes-Benz both average about €9,000 or more.

* Those brands that have the lowest average prices are Opel, Renault, and Fiat, thus their brands are generally more budget-friendly.

* Outliers: Probably due to a rather small sample or maybe the existence of high-value niche vehicles, the 'Sonstige Autos' brand has an average price as high as €12,339. Another outlier would be the 'Mini' brand, which also reflects higher average pricing in line with the brand's premium market positioning.

* Consistency: Brands like Volkswagen, Ford, and Toyota have middle-range average prices, indicative that these are models positioned to straddle affordability with premium features.

It follows, then, that this analysis suggests important ranges in the pricing and market positioning of various car brands and effectively understands the expected pricing trends for buyers and sellers.

In [45]:
premium_brand = ['audi', 'bmw', 'mercedes_benz', 'ford', 'opel', 'volkswagen']

# Counting the mean mileage
premium_mean_mileage = {}

for brand in premium_brand:
    mean_mileage = autos[autos["brand"] == brand]["odometer_km"].mean()
    premium_mean_mileage[brand] = mean_mileage.round(2)

print(premium_mean_mileage)

{'audi': 129157.39, 'bmw': 132572.51, 'mercedes_benz': 130788.36, 'ford': 124266.01, 'opel': 129310.04, 'volkswagen': 128707.16}


In [46]:
# Let`s show 6 premium brands and their mean price
premium_mean_price = {}

for brand in premium_brand:
    mean_price = autos[autos["brand"] == brand]["price"].mean()
    premium_mean_price[brand] = mean_price.round(2)

print(premium_mean_price)

{'audi': 9336.69, 'bmw': 8332.82, 'mercedes_benz': 8628.45, 'ford': 3749.47, 'opel': 2975.24, 'volkswagen': 5402.41}


In [47]:
# Convert both dictionaries to series objects, using the series constructor
bmm_series = pd.Series(premium_mean_mileage)
bmp_series = pd.Series(premium_mean_price)
print(bmm_series)
print('\n')
print(bmp_series)

audi             129157.39
bmw              132572.51
mercedes_benz    130788.36
ford             124266.01
opel             129310.04
volkswagen       128707.16
dtype: float64


audi             9336.69
bmw              8332.82
mercedes_benz    8628.45
ford             3749.47
opel             2975.24
volkswagen       5402.41
dtype: float64


In [48]:
# Create a DataFrame from the price Series and name the column 'mean_price'
df = pd.DataFrame(bmp_series, columns = ['mean_price'])

# Add a new column 'mean_mileage' to the DataFrame from the mileage Series
df['mean_mileage'] = bmm_series

# Display the DataFrame showing both average mileage and average price for each brand
df

mean_price  mean_mileage
audi              9336.69     129157.39
bmw               8332.82     132572.51
mercedes_benz     8628.45     130788.36
ford              3749.47     124266.01
opel              2975.24     129310.04
volkswagen        5402.41     128707.16

## Comparison of Average Mileage and Prices for Top 6 Car Brands
**Observations:**

1. 
Mileage and Price Trends:* 

High-end brands, including Audi, BMW, and Mercedes-Benz, have higher average prices compared to Ford, Opel, and Volkswagen. That being said, their average mileage is actually quite high, meaning that these higher brands are not necessarily used les* s.
Low-Price Brands: The average mileages of Ford and Opel are marginally lower than those of high-end brands and probably justify their lower prices as a function of lower mileage or other brand positioning strategies. The intermediate brand in terms of price and mileage is Volkswagen, probably targeting its target market with a middle road position between the high-end and low-price bran
2. ds.
Price to Mileage Analy* is:

Correlation I sight: There is no clear, direct correlation between price and mileage in this data set. It is true that some of the high-priced brands, like Audi, BMW, and Mercedes Benz, have higher mileage. However, some of the more affordable brands can have the same case. This might show that other factors may impact the price of cars aside from mi
leage.
This is an interesting comparison in informing how prices and mileages of cars are distributed across different brands that help understand the market dynamics and value proposition of each brand.

## Comparison of Average Mileage and Prices for Top 6 Car Brands
**Observations:**

*1. Mileage and Price Trends:*
* High-end brands, including Audi, BMW, and Mercedes-Benz, have higher average prices compared to Ford, Opel, and Volkswagen. That being said, their average mileage is actually quite high, meaning that these higher brands are not necessarily used less.
* Low-Price Brands: The average mileages of Ford and Opel are marginally lower than those of high-end brands and probably justify their lower prices as a function of lower mileage or other brand positioning strategies. The intermediate brand in terms of price and mileage is Volkswagen, probably targeting its target market with a middle road position between the high-end and low-price brands.

*2. Price to Mileage Analysis:*
* Correlation Insight: There is no clear, direct correlation between price and mileage in this data set. It is true that some of the high-priced brands, like Audi, BMW, and Mercedes Benz, have higher mileage. However, some of the more affordable brands can have the same case. This might show that other factors may impact the price of cars aside from mileage.

This is an interesting comparison in informing how prices and mileages of cars are distributed across different brands that help understand the market dynamics and value proposition of each brand.

In [51]:
# Display the unique values in categorical columns to identify those using German words
print(autos['seller'].unique())
print(autos['offer_type'].unique())
print(autos['abtest'].unique())
print(autos['vehicle_type'].unique())
print(autos['gear_box'].unique())
print(autos['fuel_type'].unique())
print(autos['unrepaired_damage'].unique())

['privat' 'gewerblich']
['Angebot']
['control' 'test']
['bus' 'limousine' 'kleinwagen' 'kombi' nan 'coupe' 'suv' 'cabrio'
 'andere']
['manuell' 'automatik' nan]
['lpg' 'benzin' 'diesel' nan 'cng' 'hybrid' 'elektro' 'andere']
['nein' nan 'ja']


In [52]:
# Create dictionaries for translation mappings
seller_mapping = {
    'privat': 'private',        # Mapping 'privat' to 'private'
    'gewerblich': 'commercial'  # Mapping 'gewerblich' to 'commercial'
}

offer_type_mapping = {
    'Angebot': 'offer'  # Mapping 'Angebot' to 'offer'
    # Note: 'Gesuch' (request) is not present in the dataset
}

abtest_mapping = {
    'control': 'control',  # 'control' remains 'control'
    'test': 'test'         # 'test' remains 'test'
}

vehicle_type_mapping = {
    'kleinwagen': 'small car',        # Mapping 'kleinwagen' to 'small car'
    'limousine': 'sedan',             # Mapping 'limousine' to 'sedan'
    'kombi': 'station wagon',         # Mapping 'kombi' to 'station wagon'
    'bus': 'bus',                     # 'bus' remains 'bus'
    'cabrio': 'convertible',          # Mapping 'cabrio' to 'convertible'
    'coupe': 'coupe',                 # 'coupe' remains 'coupe'
    'suv': 'suv',                     # 'suv' remains 'suv'
    'andere': 'other'                 # Mapping 'andere' to 'other'
    # Note: nan values remain unchanged
}

gearbox_mapping = {
    'manuell': 'manual',      # Mapping 'manuell' to 'manual'
    'automatik': 'automatic'  # Mapping 'automatik' to 'automatic'
    # Note: nan values remain unchanged
}

fuel_type_mapping = {
    'benzin': 'petrol',       # Mapping 'benzin' to 'petrol'
    'diesel': 'diesel',       # 'diesel' remains 'diesel'
    'lpg': 'lpg',             # 'lpg' remains 'lpg'
    'cng': 'cng',             # 'cng' remains 'cng'
    'hybrid': 'hybrid',       # 'hybrid' remains 'hybrid'
    'elektro': 'electric',    # Mapping 'elektro' to 'electric'
    'andere': 'other'         # Mapping 'andere' to 'other'
    # Note: nan values remain unchanged
}

unrepaired_damage_mapping = {
    'ja': 'yes',  # Mapping 'ja' to 'yes'
    'nein': 'no'  # Mapping 'nein' to 'no'
    # Note: nan values remain unchanged
}

# Apply the mappings to the relevant columns
autos['seller'] = autos['seller'].map(seller_mapping)  # Translate 'seller' column
autos['offer_type'] = autos['offer_type'].map(offer_type_mapping)  # Translate 'offer_type' column
autos['abtest'] = autos['abtest'].map(abtest_mapping)  # Translate 'abtest' column
autos['vehicle_type'] = autos['vehicle_type'].map(vehicle_type_mapping)  # Translate 'vehicle_type' column
autos['gear_box'] = autos['gear_box'].map(gearbox_mapping)  # Translate 'gearbox' column
autos['fuel_type'] = autos['fuel_type'].map(fuel_type_mapping)  # Translate 'fuel_type' column
autos['unrepaired_damage'] = autos['unrepaired_damage'].map(unrepaired_damage_mapping)  # Translate 'unrepaired_damage' column

# Display the transformed dataframe to check the changes
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    seller offer_type   price   abtest   vehicle_type  registration_year  \
0  private      offer  5000.0  control            bus               2004   
1  private      offer  8500.0  control          sedan               1997   
2  private      offer  8990.0     test          sedan               2009   
3  private      offer  4350.0  control      small car               2007   
4  private      offer  1350.0     test  station wagon               2003   

    gear_box  power_ps   model  odometer_km  registration_month fuel_type  \
0     manual       158  andere     150000.0                   3       lpg   
1  automatic       286     7er     150000.0                   6    petrol   
2     manual       102    golf      70000.0                   7    petrol   
3  automatic        71  fortwo      70000.0                   6    petrol   
4     manual         0   focus     150000.0                   7    petrol   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot                no  2016-03-26 00:00:00               0   
1         bmw                no  2016-04-04 00:00:00               0   
2  volkswagen                no  2016-03-26 00:00:00               0   
3       smart                no  2016-03-12 00:00:00               0   
4        ford                no  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

These translations made the dataset more readable and more useable. This transformation makes it much easier for any user, technical or non-technical, to understand the data with ease and therefore further improve analysis processes and better decision-making. Now, this dataset is clearly transformed into obvious, consistent categorical data, which clearly improves the quality and accessibility of a dataset.

In [54]:
# Convert date columns to datetime format
autos['date_crawled'] = pd.to_datetime(autos['date_crawled'])
autos['ad_created'] = pd.to_datetime(autos['ad_created'])
autos['last_seen'] = pd.to_datetime(autos['last_seen'])

# Display the transformed dataframe to check the changes
autos[['date_crawled', 'ad_created', 'last_seen']].head()

date_crawled ad_created           last_seen
0 2016-03-26 17:47:46 2016-03-26 2016-04-06 06:45:54
1 2016-04-04 13:38:56 2016-04-04 2016-04-06 14:45:08
2 2016-03-26 18:57:24 2016-03-26 2016-04-06 20:15:37
3 2016-03-12 16:58:10 2016-03-12 2016-03-15 03:16:28
4 2016-04-01 14:38:50 2016-04-01 2016-04-01 14:38:50

Such uniform numeric format eases date-related operations, makes data much more homogeneous in general, hence more manageable and analyzable for further analysis, and serves overall quality and accessibility.

In [56]:
autos.head()

date_crawled                                               name  \
0 2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1 2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2 2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3 2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4 2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    seller offer_type   price   abtest   vehicle_type  registration_year  \
0  private      offer  5000.0  control            bus               2004   
1  private      offer  8500.0  control          sedan               1997   
2  private      offer  8990.0     test          sedan               2009   
3  private      offer  4350.0  control      small car               2007   
4  private      offer  1350.0     test  station wagon               2003   

    gear_box  power_ps   model  odometer_km  registration_month fuel_type  \
0     manual       158  andere     150000.0                   3       lpg   
1  automatic       286     7er     150000.0                   6    petrol   
2     manual       102    golf      70000.0                   7    petrol   
3  automatic        71  fortwo      70000.0                   6    petrol   
4     manual         0   focus     150000.0                   7    petrol   

        brand unrepaired_damage ad_created  nr_of_pictures  postal_code  \
0     peugeot                no 2016-03-26               0        79588   
1         bmw                no 2016-04-04               0        71034   
2  volkswagen                no 2016-03-26               0        35394   
3       smart                no 2016-03-12               0        33729   
4        ford                no 2016-04-01               0        39218   

            last_seen  
0 2016-04-06 06:45:54  
1 2016-04-06 14:45:08  
2 2016-04-06 20:15:37  
3 2016-03-15 03:16:28  
4 2016-04-01 14:38:50

## Finding the Most Common Brand/Model Combinations
To understand the most prevalent brand and model combinations in our dataset, we performed the following steps:

1. Splitting the name Column: The name column, which contains both brand and model information, was split into two separate columns: brand and model. This was achieved using the str.split method with an underscore (_) as the delimiter.

2. Grouping by Brand and Model: We then grouped the dataset by the new brand and model columns using the groupby method. This allowed us to count the number of occurrences for each unique brand/model combination.

3. Counting and Sorting: After grouping, we counted the occurrences of each combination with the size method and reset the index to convert the result back to a DataFrame. We sorted this DataFrame by the count in descending order to identify the most common combinations.

4. Displaying the Results: Finally, we displayed the top 10 most common brand/model combinations.

This approach helps us identify the most popular cars in the dataset, which can provide valuable insights for market analysis and business decisions.

In [58]:
# Split the 'name' column into 'brand' and 'model' columns
autos[['brand', 'model']] = autos['name'].str.split('_', n=1, expand=True)

# Group by 'brand' and 'model' and count the occurrences
brand_model_counts = autos.groupby(['brand', 'model']).size().reset_index(name='count')

# Sort by the count in descending order to find the most common combinations
most_common_brand_model = brand_model_counts.sort_values(by='count', ascending=False).head(10)

# Display the top 10 most common brand/model combinations
print(most_common_brand_model)

            brand          model  count
31535  Volkswagen       Golf_1.4     75
4081          BMW           316i     75
9801         Ford         Fiesta     74
33133  Volkswagen           Polo     72
4321          BMW           318i     72
4685          BMW           320i     71
19953        Opel          Corsa     68
24660     Renault         Twingo     66
31508  Volkswagen           Golf     57
20018        Opel  Corsa_1.2_16V     56


This will help in getting the most popular car trends in the dataset and will be useful in gaining an insight for market analysis and business decisions.

## Price difference analysis between cars with unrepaired damage and cars without unrepaired damage
We compared average car prices, depending on whether they have unrepaired damage or not.

In [61]:
# Calculate the average price for cars with and without unrepaired damage
average_price_with_damage = autos[autos['unrepaired_damage'] == 'yes']['price'].mean()
average_price_without_damage = autos[autos['unrepaired_damage'] == 'no']['price'].mean()

# Compute the price difference
price_difference = average_price_without_damage - average_price_with_damage

# Print the results
print(f"Average price for cars with unrepaired damage: ${average_price_with_damage:.2f}")
print(f"Average price for cars without unrepaired damage: ${average_price_without_damage:.2f}")
print(f"Price difference: ${price_difference:.2f}")

Average price for cars with unrepaired damage: $2241.15
Average price for cars without unrepaired damage: $7164.03
Price difference: $4922.89


The findings reveal a significant price disparity between cars with and without unrepaired damage. On average, cars with unrepaired damage are priced approximately $4,922.89 lower than their non-damaged counterparts. This substantial difference highlights the impact of unrepaired damage on the market value of vehicles.

## Conclusion
This project tries to find out drivers of used car prices and some major trends that may exist in the dataset. It's implemented with translation of categorical values, standardization of date format, and in-depth statistical analysis of how the pricing of vehicles depends on factors like brand, model, mileage, or damage status.

These trends manifest most obviously in the price drop that cars with unrepaired damage represented: an average of `$4,922.89` less than comparable, non-damaged peers. Also noticed were dominant brand and model name combinations that shed light on the market trend and thus help stakeholders make more informed decisions.